feat(export-csv): add `Export CSV` button to download csv files

sahil143 opened the pull request:
Story: https://openshift.io/openshiftio/Openshift_io/plan/detail/1247

**Backalog Tab**
![screenshot from 2018-12-27 21-14-27](https://user-images.githubusercontent.com/9278015/50485864-ffe3ec00-0a1c-11e9-9ef9-4f3ad9901c4b.png)

**Query Tab**
![screenshot from 2018-12-27 21-14-34](https://user-images.githubusercontent.com/9278015/50485865-ffe3ec00-0a1c-11e9-92e5-18e04d641304.png)
![screenshot from 2018-12-27 21-14-48](https://user-images.githubusercontent.com/9278015/50485866-007c8280-0a1d-11e9-8397-e5d816d47c48.png)

alien-ike commented:
### Ike Plugins (test-keeper)

Thank you @sahil143 for this contribution!

<img align="left" src="https://raw.githubusercontent.com/arquillian/ike-prow-plugins/master/docs/images/arquillian_ui_failure_64px.png">

It appears that no tests have been added or updated in this PR.

Automated tests give us confidence in shipping reliable software. Please add some as part of this change.

If you are an admin or the reviewer of this PR and you are sure that no test is needed then you can use the command `/ok-without-tests` as a comment to make the status green.


For more information please head over to official [documentation](http://arquillian.org/ike-prow-plugins/#_test_keeper_plugin). You can find there how to configure the plugin.

christianvogt commented:
What happened to the formatting of the HTML?
Hard to track your changes.
Please ensure that you do not run any other formatter other than prettier.
What IDE do you use? If you use vscode, see the end of the `README.md` for extensions to install.

We should be enabling the full set of linting and formatting as soon as we can clean up the source to not have so many errors.

christianvogt commented on packages/planner/src/app/components_ngrx/planner-query/planner-query.component.less:
```diff
@@ -25,15 +25,21 @@
     top: 10px;
   }
 
+  &__export-csv {
```
Avoid using absolute layouts unless you have a very good reason to do so. Always find a way to do a flow layout. Either use flex here or responsive grid classes.

joshuawilson commented:
Please rebase.

sahil143 commented:
@joshuawilson @christianvogt I have rebased the branch and made the changes suggested by @Veethika  for `export-csv` button.
Added new screenshots in description.

vikram-raj commented on packages/planner/src/app/components_ngrx/export-csv/export-csv.component.ts:
```diff
@@ -0,0 +1,45 @@
+    this.route.queryParams.pipe(filter((q) => !!Object.keys(q).length)),
+  ).pipe(
+    switchMap(([space, query]) => {
+      console.log('################333333', query);
```
Please remove the console log.

joshuawilson commented on packages/planner/src/app/components_ngrx/export-csv/export-csv.component.ts:
```diff
@@ -0,0 +1,44 @@
+    }),
+  );
+
+  constructor(
```
The style / standard we follow has the constructor listed above the functions.

sahil143 replied:
That's an Observable being assigned to `csvLink` variable. should I move it below constructor?

joshuawilson replied:
I'm not blocking the merge on this, it just looks odd to me. It is not just a variable name that is being set. You are doing something there and that usually happens below the constructor. This is for readability not functionality.

sahil143 replied:
Done. Listed the `csvLink` below constructor.

joshuawilson commented on packages/planner/src/app/components_ngrx/export-csv/export-csv.component.ts:
```diff
@@ -0,0 +1,44 @@
+import { Component, Input, Inject } from '@angular/core';
+import { FilterService } from '../../services/filter.service';
+import { WIT_API_URL, Space } from 'ngx-fabric8-wit';
```
It appears that `space` isn't used.

joshuawilson commented:
@sahil143 I don't see approval from @fabric8-ui/uxd on the location of the button. 
This also needs unit tests.

Veethika commented:
Suggested to put the Export button on the top and move 'Show completed' and 'Show tree' to the bar below as the later is more connected to the table and should be placed closer
> 
> **Backalog Tab**
> ![screenshot from 2018-12-27 21-14-27](https://user-images.githubusercontent.com/9278015/50485864-ffe3ec00-0a1c-11e9-9ef9-4f3ad9901c4b.png)
> 

Here, the suggestion was to move the export and create workitem button in the bar below, differentiated by color from the query input bar to avoid accidental click. Placing the button at the exact same position as 'backlog tab' wasn't possible with current design as it would involve adding another bar above the input field adding a lot more to the unused spaces. This could, however, be taken care in future design changes.
> **Query Tab**
> ![screenshot from 2018-12-27 21-14-34](https://user-images.githubusercontent.com/9278015/50485865-ffe3ec00-0a1c-11e9-92e5-18e04d641304.png)
> ![screenshot from 2018-12-27 21-14-48](https://user-images.githubusercontent.com/9278015/50485866-007c8280-0a1d-11e9-8397-e5d816d47c48.png)

The implemented changes LGTM!

sahil143 commented:
I'm adding unit tests for the component.

fabric8cd commented:
:tada: This PR is included in version 4.3.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/fabric8-ui)
- [GitHub release](https://github.com/fabric8-ui/fabric8-ui/releases/tag/v4.3.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
